# 1 Construction Project - data analysis from field apps related to construction and PM works

![](https://www.letsbuild.com/wp-content/uploads/2019/10/shutterstock_1247187910.png)

### 2.1 Data Description 

* import libraries 
* read the data from the csv file 
* determine the data types in the file 
* understand the data shape 


In [1]:
import pandas as pd 
import numpy as np
import os, sys 
import sqlite3
import plotly as py
import plotly.express as px
import plotly.graph_objs as go
#import plotly.offline as offline 
#offline.init_notebook_mode(connected=True)

In [2]:
pm_forms_data=pd.read_csv('/kaggle/input/construction-and-project-management-example-data/Construction_Data_PM_Forms_All_Projects.csv')
pm_forms_data.head()

Ref                Status  \
0  F145185.4                Opened   
1  F1.495500  Open / Ongoing Works   
2  F1.495499  Open / Ongoing Works   
3  F1.495498  Open / Ongoing Works   
4  F1.495496  Open / Ongoing Works   

                                            Location  \
0  01 Daily Site Diary>Site Management>JPC Projec...   
1  02 Daily Work Plan>Site Management>JPC Project...   
2  02 Daily Work Plan>Site Management>JPC Project...   
3  02 Daily Work Plan>Site Management>JPC Project...   
4  02 Daily Work Plan>Site Management>JPC Project...   

                             Name     Created                       Type  \
0   1328 CM-SM-FRM-001 Site Diary  15/09/2020            Site Management   
1  SM-FRM-SUB-101 Daily Work Plan  15/09/2020  Subcontractor Inspections   
2  SM-FRM-SUB-101 Daily Work Plan  15/09/2020  Subcontractor Inspections   
3  SM-FRM-SUB-101 Daily Work Plan  15/09/2020  Subcontractor Inspections   
4  SM-FRM-SUB-101 Daily Work Plan  15/09/2020  Subcontractor Inspections   

  Status Changed  Open Actions  Total Actions Association  OverDue  Images  \
0     15/09/2020             0              0         NaN    False    True   
1     15/09/2020             0              0         NaN    False   False   
2     15/09/2020             0              0         NaN    False   False   
3     15/09/2020             0              0         NaN    False   False   
4     15/09/2020             0              0         NaN    False   False   

   Comments Documents  Project Report Forms Status Report Forms Group  
0     False     False     1328                Open    Site Management  
1     False     False     1328                Open      Subcontractor  
2     False     False     1328                Open      Subcontractor  
3     False     False     1328                Open      Subcontractor  
4     False     False     1328                Open      Subcontractor

Let us begin with the file named Construction Data PM Tasks.

In [3]:
pm_tasks_data=pd.read_csv('/kaggle/input/construction-and-project-management-example-data/Construction_Data_PM_Tasks_All_Projects.csv')
pm_tasks_data.head()

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description     Created  Target  \
0  task raised in incorrect location of this form...  14/09/2020     NaN   
1                                             Metsec  14/09/2020     NaN   
2  Good clear exclusion zones and access through ...  14/09/2020     NaN   
3                                           RC walls  14/09/2020     NaN   
4  block 02 working level has good housekeeping, ...  14/09/2020     NaN   

                                       Type             To Package  \
0     Safety Notice (Amber) - General Issue        Main Contractor   
1                      JPC - Progress Photo  Ceilings & Partitions   
2  Safety Notice (Green) - Good Observation        Main Contractor   
3                      JPC - Progress Photo       Precast Concrete   
4  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed Association  OverDue Images Comments Documents  \
0     14/09/2020  FormAnswer    False    NaN      NaN       NaN   
1     14/09/2020         NaN    False   True    False     False   
2     14/09/2020  FormAnswer    False   True    False     False   
3     14/09/2020         NaN    False   True    False     False   
4     14/09/2020  FormAnswer    False   True    False     False   

              Priority                         Cause  project Report Status  \
0  Behavioural Failure  JPC - Safety - Documentation     1328          Open   
1                  NaN                           NaN     1328        Closed   
2                  NaN         JPC - Safety - Access     1328        Closed   
3                  NaN                           NaN     1328        Closed   
4                  NaN  JPC - Safety - House Keeping     1328        Closed   

        Task Group  
0           Safety  
1  Site Management  
2           Safety  
3  Site Management  
4           Safety

In [4]:
pm_tasks_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12424 entries, 0 to 12423
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Ref             12424 non-null  object 
 1   Status          12424 non-null  object 
 2   Location        12424 non-null  object 
 3   Description     12424 non-null  object 
 4   Created         12424 non-null  object 
 5   Target          2568 non-null   float64
 6   Type            12424 non-null  object 
 7   To Package      11382 non-null  object 
 8   Status Changed  12424 non-null  object 
 9   Association     9483 non-null   object 
 10  OverDue         12424 non-null  bool   
 11  Images          12272 non-null  object 
 12  Comments        11902 non-null  object 
 13  Documents       11780 non-null  object 
 14  Priority        2366 non-null   object 
 15  Cause           9683 non-null   object 
 16  project         12424 non-null  int64  
 17  Report Status   12424 non-null 

In [5]:
pm_tasks_data.shape

(12424, 19)

Let's change some of the data types to make them more user friendly for analysis. 

In [6]:
pm_tasks_data['Created']=pd.to_datetime(pm_tasks_data['Created'], format='%d/%m/%Y')
pm_tasks_data.head(5)

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description    Created  Target  \
0  task raised in incorrect location of this form... 2020-09-14     NaN   
1                                             Metsec 2020-09-14     NaN   
2  Good clear exclusion zones and access through ... 2020-09-14     NaN   
3                                           RC walls 2020-09-14     NaN   
4  block 02 working level has good housekeeping, ... 2020-09-14     NaN   

                                       Type             To Package  \
0     Safety Notice (Amber) - General Issue        Main Contractor   
1                      JPC - Progress Photo  Ceilings & Partitions   
2  Safety Notice (Green) - Good Observation        Main Contractor   
3                      JPC - Progress Photo       Precast Concrete   
4  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed Association  OverDue Images Comments Documents  \
0     14/09/2020  FormAnswer    False    NaN      NaN       NaN   
1     14/09/2020         NaN    False   True    False     False   
2     14/09/2020  FormAnswer    False   True    False     False   
3     14/09/2020         NaN    False   True    False     False   
4     14/09/2020  FormAnswer    False   True    False     False   

              Priority                         Cause  project Report Status  \
0  Behavioural Failure  JPC - Safety - Documentation     1328          Open   
1                  NaN                           NaN     1328        Closed   
2                  NaN         JPC - Safety - Access     1328        Closed   
3                  NaN                           NaN     1328        Closed   
4                  NaN  JPC - Safety - House Keeping     1328        Closed   

        Task Group  
0           Safety  
1  Site Management  
2           Safety  
3  Site Management  
4           Safety

In [7]:
pm_tasks_data['Status Changed']=pd.to_datetime(pm_tasks_data['Status Changed'], format='%d/%m/%Y')
pm_tasks_data.head(5)

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description    Created  Target  \
0  task raised in incorrect location of this form... 2020-09-14     NaN   
1                                             Metsec 2020-09-14     NaN   
2  Good clear exclusion zones and access through ... 2020-09-14     NaN   
3                                           RC walls 2020-09-14     NaN   
4  block 02 working level has good housekeeping, ... 2020-09-14     NaN   

                                       Type             To Package  \
0     Safety Notice (Amber) - General Issue        Main Contractor   
1                      JPC - Progress Photo  Ceilings & Partitions   
2  Safety Notice (Green) - Good Observation        Main Contractor   
3                      JPC - Progress Photo       Precast Concrete   
4  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed Association  OverDue Images Comments Documents  \
0     2020-09-14  FormAnswer    False    NaN      NaN       NaN   
1     2020-09-14         NaN    False   True    False     False   
2     2020-09-14  FormAnswer    False   True    False     False   
3     2020-09-14         NaN    False   True    False     False   
4     2020-09-14  FormAnswer    False   True    False     False   

              Priority                         Cause  project Report Status  \
0  Behavioural Failure  JPC - Safety - Documentation     1328          Open   
1                  NaN                           NaN     1328        Closed   
2                  NaN         JPC - Safety - Access     1328        Closed   
3                  NaN                           NaN     1328        Closed   
4                  NaN  JPC - Safety - House Keeping     1328        Closed   

        Task Group  
0           Safety  
1  Site Management  
2           Safety  
3  Site Management  
4           Safety

In [8]:
pm_tasks_data['Task Group'].unique()

array(['Safety', 'Site Management', 'Quality', 'Design Team', nan],
      dtype=object)

In [9]:
print('Number of Projects: ', pm_tasks_data['project'].nunique())
print('Number of Trade Packages: ', pm_tasks_data['To Package'].nunique())
print('Number of Task Groups: ', pm_tasks_data['Task Group'].nunique())

Number of Projects:  8
Number of Trade Packages:  105
Number of Task Groups:  4


The dataset is made up of 8 construction projects, 106 different trades and each task is categorised into either 'Safety', 'Site Management', 'Quality' or  'Design Team'.

### 2.2 Data Cleaning 

* locate any missing values 
* check for duplicated items

In [10]:
#number of missing data points per column
missing_values_count = pm_tasks_data.isnull().sum()
missing_values_count[:]

Ref                   0
Status                0
Location              0
Description           0
Created               0
Target             9856
Type                  0
To Package         1042
Status Changed        0
Association        2941
OverDue               0
Images              152
Comments            522
Documents           644
Priority          10058
Cause              2741
project               0
Report Status         0
Task Group           50
dtype: int64

In [11]:
# how many total missing values do we have?
total_cells = np.product(pm_tasks_data.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
((total_missing/total_cells) * 100).round(2)

11.86

Overall, the dataset has approximatelty 11.86% of missing data. 

The columns with missing values include Target, To Package, Association, Images, Comments, Documents, Priority, Cause and Task Group. With the exception of the Target column, the rest are all columns with string values. The Priority column has over 10,000 missing values. If we were to drop these rows our dataset would be reduced significantly. Let's review the data and find a methodical approach to fill the  missing values. 

In [12]:
pm_tasks_data['Priority'].value_counts()

Behavioural Failure                                                 951
System Failure                                                      787
Medium                                                              221
High                                                                214
Best Practice                                                        83
System Failure - Deviation from RAMS / Manufacturer Instructions     31
Low                                                                  29
.                                                                    27
2 Week Look Ahead                                                    13
Low (resolve within 2 weeks)                                          3
Medium (resolve within 5 days)                                        2
High (resolve within 48 hours)                                        2
1 Week Look Ahead                                                     2
1 Month Look Ahead                                              

Behavioural Failure is the most applied in the Priority column. The tasks associated with this category relate to Safety. 

In [13]:
pm_tasks_data.loc[(pm_tasks_data['Priority'].isnull()) & (pm_tasks_data['Task Group']=='Safety'), 'Priority']='Behavioural Failure'

In [14]:
#check if change applied 
pm_tasks_data.loc[pm_tasks_data['Task Group'] == 'Safety']

Ref                Status  \
0      T1.23963030                  Open   
2       T141663.27  EHS Good Observation   
4       T141663.26  EHS Good Observation   
6       T141663.25  EHS Good Observation   
8       T141663.24  EHS Good Observation   
...            ...                   ...   
12419    T122039.5                Closed   
12420    T122039.3  EHS Good Observation   
12421    T122039.1                Closed   
12422    T120669.2                Closed   
12423    T120669.1                Closed   

                                                Location  \
0      JPC Project Management>EHS Management>01 Inspe...   
2      JPC Project Management>EHS Management>01 Inspe...   
4      JPC Project Management>EHS Management>01 Inspe...   
6      JPC Project Management>EHS Management>01 Inspe...   
8      JPC Project Management>EHS Management>01 Inspe...   
...                                                  ...   
12419                              1345 - DUB062 Project   
12420                              1345 - DUB062 Project   
12421                              1345 - DUB062 Project   
12422                              1345 - DUB062 Project   
12423                              1345 - DUB062 Project   

                                             Description    Created  Target  \
0      task raised in incorrect location of this form... 2020-09-14     NaN   
2      Good clear exclusion zones and access through ... 2020-09-14     NaN   
4      block 02 working level has good housekeeping, ... 2020-09-14     NaN   
6      operaiver using hand sanitizer points onsite. ... 2020-09-14     NaN   
8                number of operative fully pee compliant 2020-09-14     NaN   
...                                                  ...        ...     ...   
12419  Extend temp road and get IAC to install requir... 2020-03-11     NaN   
12420                              Canteen in condition  2020-03-11     NaN   
12421                        Complete drying room set up 2020-03-11     NaN   
12422  ga1 not available for some plant on site. requ... 2020-02-20     NaN   
12423                        light in toilet not working 2020-02-20     NaN   

                                           Type             To Package  \
0         Safety Notice (Amber) - General Issue        Main Contractor   
2      Safety Notice (Green) - Good Observation        Main Contractor   
4      Safety Notice (Green) - Good Observation       Precast Concrete   
6      Safety Notice (Green) - Good Observation  Ceilings & Partitions   
8      Safety Notice (Green) - Good Observation        Main Contractor   
...                                         ...                    ...   
12419     Safety Notice (Amber) - General Issue        Main Contractor   
12420  Safety Notice (Green) - Good Observation        Main Contractor   
12421     Safety Notice (Amber) - General Issue        Main Contractor   
12422     Safety Notice (Amber) - General Issue        Main Contractor   
12423     Safety Notice (Amber) - General Issue        Main Contractor   

      Status Changed Association  OverDue Images Comments Documents  \
0         2020-09-14  FormAnswer    False    NaN      NaN       NaN   
2         2020-09-14  FormAnswer    False   True    False     False   
4         2020-09-14  FormAnswer    False   True    False     False   
6         2020-09-14  FormAnswer    False   True    False     False   
8         2020-09-14  FormAnswer    False  False    False     False   
...              ...         ...      ...    ...      ...       ...   
12419     2020-03-23  FormAnswer    False   True     True     False   
12420     2020-03-11  FormAnswer    False   True    False     False   
12421     2020-03-19  FormAnswer    False   True     True     False   
12422     2020-02-28  FormAnswer    False  False     True      True   
12423     2020-02-27  FormAnswer    False   True    False     False   

                  Priority                              Cause  project  \


In [15]:
#the remaining rows can be filled with Other.
pm_tasks_data['Priority'] = pm_tasks_data['Priority'].fillna('Other')
pm_tasks_data['Priority']

0        Behavioural Failure
1                      Other
2        Behavioural Failure
3                      Other
4        Behavioural Failure
                ...         
12419         System Failure
12420    Behavioural Failure
12421         System Failure
12422    Behavioural Failure
12423    Behavioural Failure
Name: Priority, Length: 12424, dtype: object

In [16]:
pm_tasks_data['Target'].value_counts()

44011.0    75
44015.0    67
43998.0    52
43966.0    46
44070.0    43
           ..
43814.0     1
43792.0     1
43759.0     1
43760.0     1
43843.0     1
Name: Target, Length: 255, dtype: int64

In [17]:
pm_tasks_data['Target'].unique()

array([   nan, 44094., 44091., 44090., 44089., 44088., 44087., 44083.,
       44082., 44084., 44081., 44076., 44075., 44074., 44070., 44069.,
       44068., 44061., 44059., 44032., 44027., 44026., 44020., 44013.,
       44019., 44018., 44017., 44014., 44012., 44011., 44010., 44007.,
       43999., 44006., 44005., 43997., 44004., 44003., 43998., 43993.,
       43991., 43986., 43992., 43985., 43978., 43928., 43922., 43921.,
       43915., 43909., 43908., 43907., 43906., 43901., 43902., 43900.,
       43899., 43892., 43896., 43895., 43894., 43893., 43885., 43889.,
       43887., 43886., 43879., 43875., 43868., 43874., 43867., 43873.,
       43864., 43881., 43866., 43865., 43852., 43850., 43846., 43843.,
       43840., 43823., 43815., 43811., 43803., 43808., 43805., 43773.,
       43810., 43802., 43801., 43796., 43795., 43812., 43794., 43781.,
       43782., 43787., 43784., 43776., 43775., 43771., 43770., 43769.,
       43768., 43767., 43762., 43760., 43753., 43752., 43748., 43747.,
      

Given that the value 44011.0 has been assigned the most, we will use this value to fill in the missing data.

In [18]:
pm_tasks_data['Target'] = pm_tasks_data['Target'].fillna(44011.0)
pm_tasks_data['Target']

0        44011.0
1        44011.0
2        44011.0
3        44011.0
4        44011.0
          ...   
12419    44011.0
12420    44011.0
12421    44011.0
12422    44011.0
12423    44011.0
Name: Target, Length: 12424, dtype: float64

Now, let's take a look at the column named, Association. It is unclear what this column is for, however the values in this column include 'FormAnswer'. For this reason, it would seem that the data is missing because it hasn't been recorded. We will fill the missing values with 'Other'.

In [19]:
pm_tasks_data['Association'].value_counts()

FormAnswer       9281
ForwardedFrom     193
ForwardedTo         9
Name: Association, dtype: int64

In [20]:
pm_tasks_data['Association'] = pm_tasks_data['Association'].fillna('FormAnswer')
pm_tasks_data['Association']

0        FormAnswer
1        FormAnswer
2        FormAnswer
3        FormAnswer
4        FormAnswer
            ...    
12419    FormAnswer
12420    FormAnswer
12421    FormAnswer
12422    FormAnswer
12423    FormAnswer
Name: Association, Length: 12424, dtype: object

The column named, Cause also contains some missing values. They seem to have not been recorded. Existing values in this column include 'Safety documentation', ' House keeping', or 'plant and equipment'. It's fair to say that the missing values have simply not been recorded and similarly, we could fill them with
'Other'.

In [21]:
pm_tasks_data['Cause'].value_counts().head(5)

JPC - Safety - House Keeping      2039
JPC - Safety - Access             1577
JPC - Safety - Exclusion Zones     734
JPC - Quality - Workmanship        631
JPC - Safety - Storage             555
Name: Cause, dtype: int64

In [22]:
pm_tasks_data['Cause'] = pm_tasks_data['Cause'].fillna('JPC - Safety - House Keeping ')
pm_tasks_data['Cause']

0             JPC - Safety - Documentation
1            JPC - Safety - House Keeping 
2                    JPC - Safety - Access
3            JPC - Safety - House Keeping 
4             JPC - Safety - House Keeping
                       ...                
12419                JPC - Safety - Access
12420    JPC - Safety - Welfare Facilities
12421    JPC - Safety - Welfare Facilities
12422     JPC - Safety - Plant & Equipment
12423    JPC - Safety - Welfare Facilities
Name: Cause, Length: 12424, dtype: object

The column named, To Package indicates which trade contractor or department the task is associated with. By looking at the full dataset, it is clear that the missing values are for tasks associated with design. Therefore, it would be fair to say the data has not been recorded rather than it doesn't exist. We can fill the missing values with 'Design'.

In [23]:
pm_tasks_data['To Package'].value_counts()

Main Contractor                   3461
Construction Management            781
Concrete - Formwork                765
Internal Partitions & Ceilings     595
Groundworks & Drainage             500
                                  ... 
Pool Specialist                      1
External Render                      1
Architectural Steelwork              1
Civil Engineer                       1
Electrical Subcontractor             1
Name: To Package, Length: 105, dtype: int64

In [24]:
pm_tasks_data['To Package'] = pm_tasks_data['To Package'].fillna('Design')
pm_tasks_data['To Package']

0              Main Contractor
1        Ceilings & Partitions
2              Main Contractor
3             Precast Concrete
4             Precast Concrete
                 ...          
12419          Main Contractor
12420          Main Contractor
12421          Main Contractor
12422          Main Contractor
12423          Main Contractor
Name: To Package, Length: 12424, dtype: object

The columns headed 'Images', 'Comments' and 'Documents' contain Boolean data types. There's approx. 5% or less of missing data in these columns. Rather than assume what the value should be, we will fill the missing values with 'N/A'. 

In [25]:
pm_tasks_data[['Images','Comments','Documents']] = pm_tasks_data[['Images','Comments','Documents']].fillna('N/A')
pm_tasks_data[['Images','Comments','Documents']]

Images Comments Documents
0        N/A      N/A       N/A
1       True    False     False
2       True    False     False
3       True    False     False
4       True    False     False
...      ...      ...       ...
12419   True     True     False
12420   True    False     False
12421   True     True     False
12422  False     True      True
12423   True    False     False

[12424 rows x 3 columns]

Finally, the column named, Task Group has 50 missing values. This column details which sub-group the task is associated with. 

In [26]:
pm_tasks_data['Task Group'].value_counts()

Safety             8884
Site Management    1420
Design Team        1139
Quality             931
Name: Task Group, dtype: int64

Given that Safety has been assigned the most in this column, we will fill any missing values with the sub-category, "Safety". 

In [27]:
pm_tasks_data['Task Group'] = pm_tasks_data['Task Group'].fillna('Safety')
pm_tasks_data['Task Group']

0                 Safety
1        Site Management
2                 Safety
3        Site Management
4                 Safety
              ...       
12419             Safety
12420             Safety
12421             Safety
12422             Safety
12423             Safety
Name: Task Group, Length: 12424, dtype: object

Let's run the previous code again to check for missing values in the dataset.

In [28]:
# number of missing data points per column
missing_values_count = pm_tasks_data.isnull().sum()

# look at the # of missing points in each column
missing_values_count[:]

Ref               0
Status            0
Location          0
Description       0
Created           0
Target            0
Type              0
To Package        0
Status Changed    0
Association       0
OverDue           0
Images            0
Comments          0
Documents         0
Priority          0
Cause             0
project           0
Report Status     0
Task Group        0
dtype: int64

In [29]:
# how many total missing values do we have?
total_cells = np.product(pm_tasks_data.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
(total_missing/total_cells) * 100

0.0

All missing values have now been dealt with. Let's consider if there are any duplicates in the dataset.

In [30]:
#non-numerical data columns 
non_numeric_data = list(pm_tasks_data.select_dtypes(exclude=('int', 'float')).columns)
print(f'Сolumns without numeric data: {", ".join(non_numeric_data)}.')
print(len(non_numeric_data))

Сolumns without numeric data: Ref, Status, Location, Description, Created, Type, To Package, Status Changed, Association, OverDue, Images, Comments, Documents, Priority, Cause, Report Status, Task Group.
17


The following columns contain non-numeric data:- Ref, Status, Location, Description, Created, Type, To Package, Status Changed, Association, OverDue, Images, Comments, Documents, Priority, Cause, Report Status, Task Group. Due to the nature of the dataset, we can expect to find duplicate values in certain columns such as Status and Location. Let's take a closer look at the Description column to narrow down any duplicates. 

In [31]:
print('Duplication Location:', 100*(len(pm_tasks_data['Location']) - len(pm_tasks_data['Location'].drop_duplicates()))/len(pm_tasks_data['Location']))

Duplication Location: 95.21893110109465


In [32]:
print('Duplication Status:', 100*(len(pm_tasks_data['Status']) - len(pm_tasks_data['Status'].drop_duplicates()))/len(pm_tasks_data['Status']))

Duplication Status: 99.87121699935608


In [33]:
print('Duplication Description:', 100*(len(pm_tasks_data['Description']) - len(pm_tasks_data['Description'].drop_duplicates()))/len(pm_tasks_data['Description']))

Duplication Description: 17.353509336767548


In [34]:
pm_tasks_data[pm_tasks_data.duplicated(['Description'], keep=False)]

Ref                Status  \
11      T141663.21                  Open   
16      T141663.16                  Open   
17      T141663.15                  Open   
19      T141663.13                  Open   
20     T116412.196  Open / Ongoing Works   
...            ...                   ...   
12288    T135787.7                Closed   
12304   T135786.33                Closed   
12356    T135787.3                Closed   
12357    T135787.1                Closed   
12414   T122039.15  EHS Good Observation   

                                                Location  \
11     JPC Project Management>EHS Management>01 Inspe...   
16     JPC Project Management>EHS Management>01 Inspe...   
17     JPC Project Management>EHS Management>01 Inspe...   
19     JPC Project Management>EHS Management>01 Inspe...   
20     QC & BC(A)R>ITP 02 Architectural & M&E Service...   
...                                                  ...   
12288  JPC Project Management>EHS Inspections & Asses...   
12304  JPC Project Management>EHS Inspections & Asses...   
12356  JPC Project Management>EHS Inspections & Asses...   
12357  JPC Project Management>EHS Inspections & Asses...   
12414                              1345 - DUB062 Project   

                                             Description    Created   Target  \
11                                     toe board missing 2020-09-14  44011.0   
16                                         no ladder tag 2020-09-14  44011.0   
17                                         no ladder tag 2020-09-14  44011.0   
19     drinking bottles from contractors not being di... 2020-09-14  44011.0   
20                   Flatpack duct not sealed laterally. 2020-09-14  44094.0   
...                                                  ...        ...      ...   
12288                                       Housekeeping 2020-08-04  44011.0   
12304                                       Housekeeping 2020-07-27  44011.0   
12356  rutted ground outside temp compound to be made... 2020-07-09  44011.0   
12357  rutted ground outside temp compound to be made... 2020-07-09  44011.0   
12414                              Welfare in good order 2020-03-11  44011.0   

                                           Type             To Package  \
11        Safety Notice (Amber) - General Issue            Scaffolding   
16        Safety Notice (Amber) - General Issue    Temporary Electrics   
17        Safety Notice (Amber) - General Issue       Precast Concrete   
19        Safety Notice (Amber) - General Issue  Ceilings & Partitions   
20            JPC - Quality Defect (compliance)  Ceilings & Partitions   
...                                         ...                    ...   
12288     Safety Notice (Amber) - General Issue             JPC Labour   
12304     Safety Notice (Amber) - General Issue               Hoarding   
12356     Safety Notice (Amber) - General Issue        Main Contractor   
12357     Safety Notice (Amber) - General Issue             JPC Labour   
12414  Safety Notice (Green) - Good Observation        Main Contractor   

      Status Changed Association  OverDue Images Comments Documents  \
11        2020-09-14  FormAnswer    False   True    False     False   
16        2020-09-14  FormAnswer    False   True    False     False   
17        2020-09-14  FormAnswer    False   True    False     False   
19        2020-09-14  FormAnswer    False   True    False     False   
20        2020-09-14  FormAnswer    False   True     True     False   
...              ...         ...      ...    ...      ...       ...   
12288     2020-08-05  FormAnswer    False   True    False     False   
12304     2020-08-05  FormAnswer    False   True    False     False   
12356     2020-07-23  FormAnswer    False   True     True     False   
12357     2020-07-23  FormAnswer    False   True    False     False   
12414     2020-03-11  FormAnswer    False   True    False     False   

                  Priority                              Cause 

By locating the duplicated rows, it is clear that the Task Description relates to multiple packages. For example, items 16 and 17 refer to "no ladder tags" but this applies to both the Temp electrics package and Precast concrete package. Therefore, it is best to leave the duplicate description as is. 

Time to drill down into the analysis!

![](https://www.constructionexec.com/assets/site_18/images/article/050620095343.jpg?width=1280)

## 3. Data Analysis

### 3.1 Tasks by project

* filter tasks by project and then analyse 
* number of tasks by package 
* time taken to close out a task 
* tasks in order of priority 
* tasks in order of task group
* tasks in order of cause 

In [35]:
pm_tasks_data.head()

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description    Created   Target  \
0  task raised in incorrect location of this form... 2020-09-14  44011.0   
1                                             Metsec 2020-09-14  44011.0   
2  Good clear exclusion zones and access through ... 2020-09-14  44011.0   
3                                           RC walls 2020-09-14  44011.0   
4  block 02 working level has good housekeeping, ... 2020-09-14  44011.0   

                                       Type             To Package  \
0     Safety Notice (Amber) - General Issue        Main Contractor   
1                      JPC - Progress Photo  Ceilings & Partitions   
2  Safety Notice (Green) - Good Observation        Main Contractor   
3                      JPC - Progress Photo       Precast Concrete   
4  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed Association  OverDue Images Comments Documents  \
0     2020-09-14  FormAnswer    False    N/A      N/A       N/A   
1     2020-09-14  FormAnswer    False   True    False     False   
2     2020-09-14  FormAnswer    False   True    False     False   
3     2020-09-14  FormAnswer    False   True    False     False   
4     2020-09-14  FormAnswer    False   True    False     False   

              Priority                          Cause  project Report Status  \
0  Behavioural Failure   JPC - Safety - Documentation     1328          Open   
1                Other  JPC - Safety - House Keeping      1328        Closed   
2  Behavioural Failure          JPC - Safety - Access     1328        Closed   
3                Other  JPC - Safety - House Keeping      1328        Closed   
4  Behavioural Failure   JPC - Safety - House Keeping     1328        Closed   

        Task Group  
0           Safety  
1  Site Management  
2           Safety  
3  Site Management  
4           Safety

To simplify things, we will drop a few of the columns which are of less use to us. Namely, 'Target', 'Association','Overdue', 'Images', 'Comments' and 'Documents'. 

In [36]:
subset_project_tasks = pm_tasks_data[['Ref', 'Status', 'Location', 'Description', 'Created', 'Type', 'To Package', 'Status Changed', 'Priority', 'Cause', 'project', 'Report Status', 'Task Group']]
subset_project_tasks.dtypes

Ref                       object
Status                    object
Location                  object
Description               object
Created           datetime64[ns]
Type                      object
To Package                object
Status Changed    datetime64[ns]
Priority                  object
Cause                     object
project                    int64
Report Status             object
Task Group                object
dtype: object

In [37]:
subset_project_tasks.shape

(12424, 13)

In [38]:
#filter tasks by project and then analyse 
data_by_project = subset_project_tasks.drop(columns=['Status', 'Location', 'Description', 'Created', 'Type', 'To Package', 'Status Changed', 'Priority', 'Cause', 'Report Status', 'Task Group'])

In [39]:
data_by_project = data_by_project.groupby(by = 'project').count().rename(columns = {'Ref' : 'Quantity'}).reset_index()
data_by_project

project  Quantity
0     1328      3751
1     1329       478
2     1330      3684
3     1335      1267
4     1338      1308
5     1340       995
6     1343       381
7     1345       560

In [40]:
trace1 = go.Pie(labels = data_by_project['project'],
             values = data_by_project['Quantity'],
              
            title = 'Number of tasks per project'),
           
fig = go.Figure(trace1)
fig.update_layout(legend_title_text = 'Project Ref.') 
fig.show()

### 3.2 Tasks by package

In [41]:
#filter tasks by trade package
subset_project_tasks.head()

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description    Created  \
0  task raised in incorrect location of this form... 2020-09-14   
1                                             Metsec 2020-09-14   
2  Good clear exclusion zones and access through ... 2020-09-14   
3                                           RC walls 2020-09-14   
4  block 02 working level has good housekeeping, ... 2020-09-14   

                                       Type             To Package  \
0     Safety Notice (Amber) - General Issue        Main Contractor   
1                      JPC - Progress Photo  Ceilings & Partitions   
2  Safety Notice (Green) - Good Observation        Main Contractor   
3                      JPC - Progress Photo       Precast Concrete   
4  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed             Priority                          Cause  project  \
0     2020-09-14  Behavioural Failure   JPC - Safety - Documentation     1328   
1     2020-09-14                Other  JPC - Safety - House Keeping      1328   
2     2020-09-14  Behavioural Failure          JPC - Safety - Access     1328   
3     2020-09-14                Other  JPC - Safety - House Keeping      1328   
4     2020-09-14  Behavioural Failure   JPC - Safety - House Keeping     1328   

  Report Status       Task Group  
0          Open           Safety  
1        Closed  Site Management  
2        Closed           Safety  
3        Closed  Site Management  
4        Closed           Safety

In [42]:
package_df = subset_project_tasks['To Package'].value_counts().reset_index().rename(columns={'index':'Trade Package', 'To Package': 'Quantity'})
#package_df = package_df.reset_index()
package_df.head(10)

Trade Package  Quantity
0                 Main Contractor      3461
1                          Design      1042
2         Construction Management       781
3             Concrete - Formwork       765
4  Internal Partitions & Ceilings       595
5          Groundworks & Drainage       500
6                        Formwork       390
7                        Concrete       331
8                     Scaffolding       330
9                    Ground Works       328

In [43]:
trace2 = go.Bar(x=package_df['Quantity'],
            y=package_df['Trade Package'].head(10),
            orientation='h',
            text=package_df['Quantity'],
            textposition='inside')
fig = go.Figure(trace2)
fig = fig.update_layout(title = 'Number of tasks per trade package', xaxis_title='Quantity',yaxis_title='Trade Package', xaxis={'categoryorder': 'total descending'})
fig.show()

### 3.3 Average time taken to close out a task by project

In [44]:
subset_project_tasks.dtypes

Ref                       object
Status                    object
Location                  object
Description               object
Created           datetime64[ns]
Type                      object
To Package                object
Status Changed    datetime64[ns]
Priority                  object
Cause                     object
project                    int64
Report Status             object
Task Group                object
dtype: object

In [45]:
close_out_time = subset_project_tasks[['Ref', 'Status', 'Location', 'Description', 'Created', 'Status Changed', 'project', 'Task Group']]
close_out_time.head()

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                            Location  \
0  JPC Project Management>EHS Management>01 Inspe...   
1  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2  JPC Project Management>EHS Management>01 Inspe...   
3  QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4  JPC Project Management>EHS Management>01 Inspe...   

                                         Description    Created  \
0  task raised in incorrect location of this form... 2020-09-14   
1                                             Metsec 2020-09-14   
2  Good clear exclusion zones and access through ... 2020-09-14   
3                                           RC walls 2020-09-14   
4  block 02 working level has good housekeeping, ... 2020-09-14   

  Status Changed  project       Task Group  
0     2020-09-14     1328           Safety  
1     2020-09-14     1328  Site Management  
2     2020-09-14     1328           Safety  
3     2020-09-14     1328  Site Management  
4     2020-09-14     1328           Safety

In [46]:
date_completed = close_out_time['Status Changed']
date_created = close_out_time['Created']

close_out_time['Time to complete task'] = date_completed - date_created
close_out_time

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Ref                Status  \
0      T1.23963030                  Open   
1      T116412.200                Closed   
2       T141663.27  EHS Good Observation   
3      T116412.199                Closed   
4       T141663.26  EHS Good Observation   
...            ...                   ...   
12419    T122039.5                Closed   
12420    T122039.3  EHS Good Observation   
12421    T122039.1                Closed   
12422    T120669.2                Closed   
12423    T120669.1                Closed   

                                                Location  \
0      JPC Project Management>EHS Management>01 Inspe...   
1      QC & BC(A)R>ITP 02 Architectural & M&E Service...   
2      JPC Project Management>EHS Management>01 Inspe...   
3      QC & BC(A)R>ITP 02 Architectural & M&E Service...   
4      JPC Project Management>EHS Management>01 Inspe...   
...                                                  ...   
12419                              1345 - DUB062 Project   
12420                              1345 - DUB062 Project   
12421                              1345 - DUB062 Project   
12422                              1345 - DUB062 Project   
12423                              1345 - DUB062 Project   

                                             Description    Created  \
0      task raised in incorrect location of this form... 2020-09-14   
1                                                 Metsec 2020-09-14   
2      Good clear exclusion zones and access through ... 2020-09-14   
3                                               RC walls 2020-09-14   
4      block 02 working level has good housekeeping, ... 2020-09-14   
...                                                  ...        ...   
12419  Extend temp road and get IAC to install requir... 2020-03-11   
12420                              Canteen in condition  2020-03-11   
12421                        Complete drying room set up 2020-03-11   
12422  ga1 not available for some plant on site. requ... 2020-02-20   
12423                        light in toilet not working 2020-02-20   

      Status Changed  project       Task Group Time to complete task  
0         2020-09-14     1328           Safety                0 days  
1         2020-09-14     1328  Site Management                0 days  
2         2020-09-14     1328           Safety                0 days  
3         2020-09-14     1328  Site Management                0 days  
4         2020-09-14     1328           Safety                0 days  
...              ...      ...              ...                   ...  
12419     2020-03-23     1345           Safety               12 days  
12420     2020-03-11     1345           Safety                0 days  
12421     2020-03-19     1345           Safety                8 days  
12422     2020-02-28     1345           Safety                8 days  
12423     2020-02-27     1345           Safety                7 days  

[12424 rows x 9 columns]

In [47]:
avg_close_out_time=close_out_time.groupby(['project','Task Group'])[['Time to complete task']].mean().sort_values('Time to complete task', ascending=False).reset_index()
avg_close_out_time['Time to complete task']=avg_close_out_time['Time to complete task'].dt.days
avg_close_out_time

project       Task Group  Time to complete task
0      1328      Design Team                     77
1      1335      Design Team                     67
2      1328          Quality                     31
3      1330      Design Team                     29
4      1328  Site Management                     24
5      1330          Quality                     19
6      1330  Site Management                     19
7      1329  Site Management                     18
8      1340      Design Team                     17
9      1335          Quality                     13
10     1338      Design Team                     10
11     1343  Site Management                      9
12     1338           Safety                      9
13     1330           Safety                      7
14     1340          Quality                      7
15     1335  Site Management                      6
16     1329           Safety                      5
17     1335           Safety                      5
18     1328           Safety                      4
19     1345           Safety                      3
20     1343           Safety                      3
21     1338          Quality                      2
22     1338  Site Management                      2
23     1340           Safety                      2
24     1329          Quality                      0
25     1340  Site Management                      0
26     1345          Quality                      0
27     1329      Design Team                      0
28     1343      Design Team                      0
29     1345      Design Team                      0
30     1345  Site Management                      0

In [48]:
avg_close_out_time['project'].unique()

array([1328, 1335, 1330, 1329, 1340, 1338, 1343, 1345])

In [49]:
fig=px.bar(avg_close_out_time, x='project', y='Time to complete task', color='Task Group', title='Time taken to complete tasks by project and department')
#fig.layout = dict(xaxis=dict(type="category"))
fig.show()

In [50]:
#central trend of time to compltete a task 
close_out_time.describe().T

count                        mean  \
project                12424.0                  1332.58548   
Time to complete task    12424  11 days 01:20:12.363168061   

                                              std              min  \
project                                  5.213831           1328.0   
Time to complete task  28 days 05:54:51.138586493  0 days 00:00:00   

                                   25%              50%              75%  \
project                         1328.0           1330.0           1338.0   
Time to complete task  0 days 00:00:00  1 days 00:00:00  8 days 00:00:00   

                                     max  
project                           1345.0  
Time to complete task  317 days 00:00:00

In [51]:
#On average, it took 11 days to close out a task, with the longest taking 598 days. There is a minimum of -304 days,
#which suggests there could be some sort of input error. 

On average, it took 11 days to close out a task, with the longest task duration taking 598 days. There is a minimum of -304 days,
#which suggests there could be some sort of input error.

In [52]:
close_out_time.dtypes

Ref                               object
Status                            object
Location                          object
Description                       object
Created                   datetime64[ns]
Status Changed            datetime64[ns]
project                            int64
Task Group                        object
Time to complete task    timedelta64[ns]
dtype: object

In [53]:
pd.options.display.max_colwidth = 100

In [54]:
close_out_time.loc[close_out_time['Time to complete task'] > '11 days', :].sort_values(by='Time to complete task', ascending=False)

Ref                       Status  \
3733      T85112.5  3rd Party Sign Off - Closed   
2427     T92999.39                 JPC Sign Off   
3467   T1.16048246  3rd Party Sign Off - Closed   
3732      T85112.6  3rd Party Sign Off - Closed   
2424     T92999.42  3rd Party Sign Off - Closed   
...            ...                          ...   
4429   T102314.116                       Closed   
10149  T1.22239774                       Closed   
10150  T1.22239773                       Closed   
10151  T1.22239771                       Closed   
12419    T122039.5                       Closed   

                                                                     Location  \
3733                  QC & BC(A)R>ITP 01 Structural>Block 3>Ground Floor Plan   
2427                        QC & BC(A)R>ITP 01 Structural>Block 1>Basement GA   
3467                  QC & BC(A)R>ITP 01 Structural>Block 3>Ground Floor Plan   
3732                  QC & BC(A)R>ITP 01 Structural>Block 3>Ground Floor Plan   
2424   QC & BC(A)R>ITP 02 Architectural & M&E Services>Block 3>1st Floor Plan   
...                                                                       ...   
4429                     JPC Project Management>EHS Management>01 Inspections   
10149                                                  1338 - Kildare Village   
10150                                                  1338 - Kildare Village   
10151                                                  1338 - Kildare Village   
12419                                                   1345 - DUB062 Project   

                                                                                               Description  \
3733                   Steel reinforcement certification to forwarded to structural engineer for approval.   
2427   Task - Block 1; Backfilling in progress, material to be SR21 compliant and compact as per C&S sp...   
3467   Beam on GL 5\nIn one area at GL E end of the beam, there is one bar which is touching the side o...   
3732         Protection of adjoining buildings. Method of preventing interference with adjoining property.   
2424   Task Block 3 - honeycombing in concrete evident in some locations. concrete to vibrated sufficie...   
...                                                                                                    ...   
4429                                   Some exposed rebar along mezzanine slab and lift shaft at 1st Floor   
10149                                                                                           Demolition   
10150                                                                                Removing Sheet Piling   
10151                                                                                             Drainage   
12419                                            Extend temp road and get IAC to install required walkways   

         Created Status Changed  project       Task Group  \
3733  2019-05-20     2020-04-01     1328      Design Team   
2427  2019-11-19     2020-09-03     1328      Design Team   
3467  2019-07-22     2020-04-27     1328      Design Team   
3732  2019-05-20     2020-02-24     1328      Design Team   
2424  2019-11-19     2020-08-25     1328      Design Team   
...          ...            ...      ...              ...   
4429  2020-07-23     2020-08-04     1330           Safety   
10149 2020-06-11     2020-06-23     1338  Site Management   
10150 2020-06-11     2020-06-23     1338  Site Management   
10151 2020-06-11     2020-06-23     1338  Site Management   
12419 2020-03-11     2020-03-23     1345           Safety   

      Time to complete task  
3733               317 days  
2427               289 days  
3467               280 days  
3732               280 days  
2424               280 days  
...                     ...  
4429                12 days  
10149               12 days  
10150               12 days  
10151               12 days  
12419               12 days  

[2518 rows x 9

In [55]:
close_out_time['Time to complete task'].value_counts()

0 days      5949
1 days       707
7 days       496
6 days       477
2 days       441
            ... 
232 days       1
271 days       1
279 days       1
317 days       1
188 days       1
Name: Time to complete task, Length: 193, dtype: int64

In [56]:
#let's have a look at the average time taken to complete a task by each task group
time_by_taskgrp_df = close_out_time
#time_by_taskgrp_df.dtypes
time_by_taskgrp_df = time_by_taskgrp_df.groupby('Task Group')['Time to complete task'].mean(numeric_only=False).reset_index()
time_by_taskgrp_df


Task Group      Time to complete task
0      Design Team 41 days 06:05:22.388059701
1          Quality 22 days 10:51:10.247046187
2           Safety  5 days 12:44:48.515782404
3  Site Management 14 days 02:52:23.661971831

In [57]:
#The table above provides good insight into the areas where construction projects need more efficiency. It is not
#very surprising for design tasks to take longer, but perhaps the average of 40days could be improved on. Also, 
#safety and quality are key areas where further improvements could be made. 

In [58]:
#4. tasks in order of priority 
priority_df = subset_project_tasks
priority_df.head()

Ref                Status  \
0  T1.23963030                  Open   
1  T116412.200                Closed   
2   T141663.27  EHS Good Observation   
3  T116412.199                Closed   
4   T141663.26  EHS Good Observation   

                                                                                   Location  \
0                                      JPC Project Management>EHS Management>01 Inspections   
1  QC & BC(A)R>ITP 02 Architectural & M&E Services>Block 1>Block 1 - Bracken Lane Elevation   
2                                      JPC Project Management>EHS Management>01 Inspections   
3                    QC & BC(A)R>ITP 02 Architectural & M&E Services>Block 2>5th Floor Plan   
4                                      JPC Project Management>EHS Management>01 Inspections   

                                                                                           Description  \
0  task raised in incorrect location of this form. see field "Have the EHS task(s), raised from the...   
1                                                                                               Metsec   
2                                                   Good clear exclusion zones and access through site   
3                                                                                             RC walls   
4                 block 02 working level has good housekeeping, clear access, and good edge protection   

     Created                                      Type             To Package  \
0 2020-09-14     Safety Notice (Amber) - General Issue        Main Contractor   
1 2020-09-14                      JPC - Progress Photo  Ceilings & Partitions   
2 2020-09-14  Safety Notice (Green) - Good Observation        Main Contractor   
3 2020-09-14                      JPC - Progress Photo       Precast Concrete   
4 2020-09-14  Safety Notice (Green) - Good Observation       Precast Concrete   

  Status Changed             Priority                          Cause  project  \
0     2020-09-14  Behavioural Failure   JPC - Safety - Documentation     1328   
1     2020-09-14                Other  JPC - Safety - House Keeping      1328   
2     2020-09-14  Behavioural Failure          JPC - Safety - Access     1328   
3     2020-09-14                Other  JPC - Safety - House Keeping      1328   
4     2020-09-14  Behavioural Failure   JPC - Safety - House Keeping     1328   

  Report Status       Task Group  
0          Open           Safety  
1        Closed  Site Management  
2        Closed           Safety  
3        Closed  Site Management  
4        Closed           Safety

In [59]:
trace4 = go.Pie(labels = priority_df['Priority'],
             values = priority_df['Priority'].value_counts(),
              
            title = 'Tasks split by priority'),
           
fig = go.Figure(trace4)
fig.show()

In [60]:
#The majority of tasks are allocated to behavioural failure. Let's get a better idea of the task description for 
#these items.
priority_df.loc[priority_df['Priority']=='Behavioural Failure', :]

Ref                Status  \
0      T1.23963030                  Open   
2       T141663.27  EHS Good Observation   
4       T141663.26  EHS Good Observation   
6       T141663.25  EHS Good Observation   
8       T141663.24  EHS Good Observation   
...            ...                   ...   
12416   T122039.11  EHS Good Observation   
12418    T122039.7  EHS Good Observation   
12420    T122039.3  EHS Good Observation   
12422    T120669.2                Closed   
12423    T120669.1                Closed   

                                                   Location  \
0      JPC Project Management>EHS Management>01 Inspections   
2      JPC Project Management>EHS Management>01 Inspections   
4      JPC Project Management>EHS Management>01 Inspections   
6      JPC Project Management>EHS Management>01 Inspections   
8      JPC Project Management>EHS Management>01 Inspections   
...                                                     ...   
12416                                 1345 - DUB062 Project   
12418                                 1345 - DUB062 Project   
12420                                 1345 - DUB062 Project   
12422                                 1345 - DUB062 Project   
12423                                 1345 - DUB062 Project   

                                                                                               Description  \
0      task raised in incorrect location of this form. see field "Have the EHS task(s), raised from the...   
2                                                       Good clear exclusion zones and access through site   
4                     block 02 working level has good housekeeping, clear access, and good edge protection   
6                                              operaiver using hand sanitizer points onsite. Good practice   
8                                                                  number of operative fully pee compliant   
...                                                                                                    ...   
12416                                               Good management of exclusion zones to overhead cables    
12418                                                     Exclusion zone to overhead lines well maintained   
12420                                                                                Canteen in condition    
12422                        ga1 not available for some plant on site. requested from FIMA Safety Officer.   
12423                                                                          light in toilet not working   

         Created                                      Type  \
0     2020-09-14     Safety Notice (Amber) - General Issue   
2     2020-09-14  Safety Notice (Green) - Good Observation   
4     2020-09-14  Safety Notice (Green) - Good Observation   
6     2020-09-14  Safety Notice (Green) - Good Observation   
8     2020-09-14  Safety Notice (Green) - Good Observation   
...          ...                                       ...   
12416 2020-03-11  Safety Notice (Green) - Good Observation   
12418 2020-03-11  Safety Notice (Green) - Good Observation   
12420 2020-03-11  Safety Notice (Green) - Good Observation   
12422 2020-02-20     Safety Notice (Amber) - General Issue   
12423 2020-02-20     Safety Notice (Amber) - General Issue   

                  To Package Status Changed             Priority  \
0            Main Contractor     2020-09-14  Behavioural Failure   
2            Main Contractor     2020-09-14  Behavioural Failure   
4           Precast Concrete     2020-09-14  Behavioural Failure   
6      Ceilings & Partitions     2020-09-14  Behavioural Failure   
8            Main Contractor     2020-09-14  Behavioural Failure   
...                      ...            ...                  ...   
12416        Main Contractor     2020-03-11  Behavioural Failure   
12418        Main Contractor     2020-03-11  Behavioural Failure   
12420        Main Contractor     2020-03-11  Behavi

In [61]:
#It seems that these particular tasks are general housekeeping and health & safety tasks, and they are also 
#categorised under the 'safety' task group. 

In [62]:
#5. tasks in order of task group
task_group_df = subset_project_tasks
task_group_df['Task Group'].unique()

array(['Safety', 'Site Management', 'Quality', 'Design Team'],
      dtype=object)

In [63]:
fig=px.histogram(task_group_df, x='Task Group', title='Task Group Histogram')
fig.show()

In [64]:
#Due to the nature of construction projects and constantly having to adapt and adjust, we can expect saftey tasks
#to take up a lot of time and resources. Let's take a closer look at the nature of safety tasks.

In [65]:
task_group_df.loc[task_group_df['Task Group']=='Safety', :]

Ref                Status  \
0      T1.23963030                  Open   
2       T141663.27  EHS Good Observation   
4       T141663.26  EHS Good Observation   
6       T141663.25  EHS Good Observation   
8       T141663.24  EHS Good Observation   
...            ...                   ...   
12419    T122039.5                Closed   
12420    T122039.3  EHS Good Observation   
12421    T122039.1                Closed   
12422    T120669.2                Closed   
12423    T120669.1                Closed   

                                                   Location  \
0      JPC Project Management>EHS Management>01 Inspections   
2      JPC Project Management>EHS Management>01 Inspections   
4      JPC Project Management>EHS Management>01 Inspections   
6      JPC Project Management>EHS Management>01 Inspections   
8      JPC Project Management>EHS Management>01 Inspections   
...                                                     ...   
12419                                 1345 - DUB062 Project   
12420                                 1345 - DUB062 Project   
12421                                 1345 - DUB062 Project   
12422                                 1345 - DUB062 Project   
12423                                 1345 - DUB062 Project   

                                                                                               Description  \
0      task raised in incorrect location of this form. see field "Have the EHS task(s), raised from the...   
2                                                       Good clear exclusion zones and access through site   
4                     block 02 working level has good housekeeping, clear access, and good edge protection   
6                                              operaiver using hand sanitizer points onsite. Good practice   
8                                                                  number of operative fully pee compliant   
...                                                                                                    ...   
12419                                            Extend temp road and get IAC to install required walkways   
12420                                                                                Canteen in condition    
12421                                                                          Complete drying room set up   
12422                        ga1 not available for some plant on site. requested from FIMA Safety Officer.   
12423                                                                          light in toilet not working   

         Created                                      Type  \
0     2020-09-14     Safety Notice (Amber) - General Issue   
2     2020-09-14  Safety Notice (Green) - Good Observation   
4     2020-09-14  Safety Notice (Green) - Good Observation   
6     2020-09-14  Safety Notice (Green) - Good Observation   
8     2020-09-14  Safety Notice (Green) - Good Observation   
...          ...                                       ...   
12419 2020-03-11     Safety Notice (Amber) - General Issue   
12420 2020-03-11  Safety Notice (Green) - Good Observation   
12421 2020-03-11     Safety Notice (Amber) - General Issue   
12422 2020-02-20     Safety Notice (Amber) - General Issue   
12423 2020-02-20     Safety Notice (Amber) - General Issue   

                  To Package Status Changed             Priority  \
0            Main Contractor     2020-09-14  Behavioural Failure   
2            Main Contractor     2020-09-14  Behavioural Failure   
4           Precast Concrete     2020-09-14  Behavioural Failure   
6      Ceilings & Partitions     2020-09-14  Behavioural Failure   
8            Main Contractor     2020-09-14  Behavioural Failure   
...                      ...            ...                  ...   
12419        Main Contractor     2020-03-23       System Failure   
12420        Main Contractor     2020-03-11  Behavioural Failure   
12421        Main Contractor     2020-03-19       S

In [66]:
#From the descriptions above, safety tasks are quite a wide range. Safety tasks appear frequently, due to the 
#changing nature of construction sites. 

In [67]:
#6. tasks in order of cause 

In [68]:
cause_df = subset_project_tasks
cause_df['Cause'].unique()

array(['JPC - Safety - Documentation', 'JPC - Safety - House Keeping ',
       'JPC - Safety - Access', 'JPC - Safety - House Keeping',
       'JPC - Safety - Welfare Facilities', 'JPC - Safety - PPE',
       'JPC - Safety - Storage', 'JPC - Safety - Ladders',
       'JPC - Safety - WAH Equipment', 'JPC - Safety - Scaffolding',
       'JPC - Quality - Workmanship',
       'JPC - Safety - Temporary Electrics', 'JPC - Safety - Electrical',
       'JPC - Safety - Lead Management',
       'JPC - Safety - Working At Height',
       'JPC - Program - 6 Information', 'JPC - Program - 5 Weather',
       'JPC - Safety - Demolition', 'JPC - Safety - Plant & Equipment',
       'JPC - Safety - Environmental', 'JPC - Quality - Materials',
       'JPC - Safety - Excavations', 'JPC - Quality - Management',
       'JPC - Quality - Damage', 'JPC - Quality - Information',
       'JPC - Safety - Exclusion Zones', 'JPC - Program - 1 Materials',
       'JPC - Safety - Lifting / Slinging',
       'JPC - Safe

In [69]:
#The cause or reason column has a wide range of descriptions. Within the column, there are sub-categories, including:
# - safety
# - quality
# - program
# - communication 
# Further, within the sub-categories there is more specific detail such as variation, client-driven, damage etc.

In [70]:
#Due to the wide-range of detail, let's look at the top 10. 

In [71]:
#dfg = df.groupby(['name']).size().to_frame().sort_values([0], ascending = False).head(10).reset_index()
#fig = px.histogram(dfg, x='name', y = 'count')

cause_df = cause_df.groupby(['Cause']).size().to_frame('Quantity').sort_values(by='Quantity', ascending=False).head(10).reset_index()
cause_df

Cause  Quantity
0      JPC - Safety - House Keeping       2741
1       JPC - Safety - House Keeping      2039
2              JPC - Safety - Access      1577
3     JPC - Safety - Exclusion Zones       734
4        JPC - Quality - Workmanship       631
5             JPC - Safety - Storage       555
6          JPC - Safety - Management       384
7       JPC - Safety - Environmental       311
8       JPC - Safety - Documentation       291
9  JPC - Safety - Welfare Facilities       271

In [72]:
fig=px.histogram(cause_df, x='Cause', y='Quantity', title='Cause Histogram')
fig.show()

In [73]:
#With so many tasks listed under 'Other', perhaps there is scope to add more sub-categories to be more specific. 
#Areas for improvement which fall under the 'Safety' category, are house keeping, access and exclusion zones. 


In [74]:
#format data in SQL datbase 

In [75]:
#select columns from pandas df to be used in a SQL database 
cols_to_use=pm_tasks_data[['Created', 'To Package', 'Status Changed', 'project', 'Task Group']]
cols_to_use

Created             To Package Status Changed  project  \
0     2020-09-14        Main Contractor     2020-09-14     1328   
1     2020-09-14  Ceilings & Partitions     2020-09-14     1328   
2     2020-09-14        Main Contractor     2020-09-14     1328   
3     2020-09-14       Precast Concrete     2020-09-14     1328   
4     2020-09-14       Precast Concrete     2020-09-14     1328   
...          ...                    ...            ...      ...   
12419 2020-03-11        Main Contractor     2020-03-23     1345   
12420 2020-03-11        Main Contractor     2020-03-11     1345   
12421 2020-03-11        Main Contractor     2020-03-19     1345   
12422 2020-02-20        Main Contractor     2020-02-28     1345   
12423 2020-02-20        Main Contractor     2020-02-27     1345   

            Task Group  
0               Safety  
1      Site Management  
2               Safety  
3      Site Management  
4               Safety  
...                ...  
12419           Safety  
12420           Safety  
12421           Safety  
12422           Safety  
12423           Safety  

[12424 rows x 5 columns]

In [76]:
DATABASE_FILEPATH='const_project_data.db'

if os.path.exists(DATABASE_FILEPATH):
    os.remove(DATABASE_FILEPATH)
    
conn = sqlite3.connect(DATABASE_FILEPATH)
cur = conn.cursor()

In [77]:
cur.execute('''

CREATE TABLE IF NOT EXISTS project_tasks (

project INTEGER, 
Created DATE, 
StatusChanged DATE, 
ToPackage VARCHAR, 
TaskGroup VARCHAR)

 ''')
conn.commit()

df=cols_to_use
df.to_sql('project_tasks', conn, if_exists = 'replace', index = False)

cur.execute('''
SELECT * FROM project_tasks
''')

/opt/conda/lib/python3.7/site-packages/pandas/core/generic.py:2882: UserWarning:

The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.



In [78]:
for row in cur.fetchall():
    print(row)

('2020-09-14 00:00:00', 'Main Contractor', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Ceilings & Partitions', '2020-09-14 00:00:00', 1328, 'Site Management')
('2020-09-14 00:00:00', 'Main Contractor', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Precast Concrete', '2020-09-14 00:00:00', 1328, 'Site Management')
('2020-09-14 00:00:00', 'Precast Concrete', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Roofing', '2020-09-14 00:00:00', 1328, 'Site Management')
('2020-09-14 00:00:00', 'Ceilings & Partitions', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Cladding & Roofing', '2020-09-14 00:00:00', 1328, 'Site Management')
('2020-09-14 00:00:00', 'Main Contractor', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Main Contractor', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Ceilings & Partitions', '2020-09-14 00:00:00', 1328, 'Safety')
('2020-09-14 00:00:00', 'Scaffolding', '2020-09